runtime/strategies/coalesce-recipes.js

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2018 Google Inc. All rights reserved.
+// This code may only be used under the BSD style license found at
+// http://polymer.github.io/LICENSE.txt
+// Code distributed by Google as part of this project is also
+// subject to an additional IP rights grant found at
+// http://polymer.github.io/PATENTS.txt
+
+import {Strategy} from '../../strategizer/strategizer.js';
+import Recipe from '../recipe/recipe.js';
+import RecipeUtil from '../recipe/recipe-util.js';
+import RecipeWalker from '../recipe/walker.js';
+import Handle from '../recipe/handle.js';
+
+// This strategy coalesces unresolved terminal recipes (i.e. those that cannot
+// be modified by any strategy apart from this one) by finding unresolved
+// use/map/copy handles from 2 recipes and replacing them with a single create
+// handle that allows to satisfy all constraints.
+export default class CoalesceRecipes extends Strategy {
+
+  getResults(inputParams) {
+    return inputParams.terminal.filter(result => !result.result.isResolved());
+  }
+
+  async generate(inputParams) {
+    let self = this;
+    let applicableHandles = null;
+    return Recipe.over(this.getResults(inputParams), new class extends RecipeWalker {
+      onHandle(recipe, handle) {
+        if (handle.fate === 'create'
+            || handle.id
+            || handle.connections.length === 0) return;
+
+        let counts = RecipeUtil.directionCounts(handle);
+
+        if (applicableHandles === null)
+          applicableHandles = self.findApplicableHandles(inputParams);
+
+        let results = [];
+
+        for (let [otherHandle, otherCounts] of applicableHandles) {
+          if (otherHandle.recipe === recipe) continue;
+          if (otherCounts.in + counts.in === 0
```

Comment:
why is this?

Reply:
The idea is that we want to connect an unresolvable writing connection to unresolvable reading, or the other way around. Connection where both particles only write or only read are not considered useful here - though this can be debated.